diff --git a/caches/run_test.py b/caches/run_test.py
index 3d90f81..11c7776 100755
--- a/caches/run_test.py
+++ b/caches/run_test.py
@@ -50,8 +50,6 @@ variables = (
 NOPLOT = False
 try:
     import matplotlib
-    matplotlib.use('PDF') # must be called immediately, and before import pylab
-                          # sets the back-end for matplotlib
     import matplotlib.pyplot as plt
     import numpy as np
     import pylab
@@ -90,7 +88,7 @@ def outputAccessLatency(data, size, memstring, freq):
 # 1. Parses input file.
 # 2. Runs benchmark (many times) and records results in results file.
 # 3. Parses results file for data (can also play back recording using NORUN option).
-# 4. Graphs results and outputs to .pdf file.
+# 4. Graphs results and outputs to .png file.
 def main():
     # handle CLI options
     ccbench.controller()  
@@ -197,7 +195,7 @@ def main():
     plt.xticks(
         xtick_range,
         xtick_names,
-        rotation='-30',
+        rotation=-30,
         size='small' 
         )
     plt.yticks(ytick_range,ytick_names)
@@ -285,7 +283,7 @@ def main():
         
     plt.savefig(filename)
     print(("Used report filename             : " + report_filename)) 
-    print(("Finished Plotting, saved as file : " + filename + ".pdf"))
+    print(("Finished Plotting, saved as file : " + filename + ".png"))
 
 
 
